multi: query chan update timestamps

ellemouton opened the pull request:
In this PR, we start supporting the timestamps query option in the `query_channel_range` message. If we receive such a message with that bit set then we now populate the timestamps field of the `reply_channel_range` message. 

With this feature we can now ask for the timestamps for the latest channel updates for a channel to be sent along with the SCID. Thus, if we have marked one of the channels in the list as a zombie but we see that the timestamp sent with it make it not a zombie, then we request the channel_update messages for that channel so that we can perhaps awaken the channel. 

### How to test?

See this comment: https://github.com/lightningnetwork/lnd/pull/8030#issuecomment-1808145257

Fixes https://github.com/lightningnetwork/lnd/issues/7989

Roasbeef commented on discovery/gossiper.go:
```diff
@@ -329,6 +329,11 @@ type Config struct {
 	FindChannel func(node *btcec.PublicKey, chanID lnwire.ChannelID) (
 		*channeldb.OpenChannel, error)
+
+	// IsZombieChannel takes the timestamps of the latest channel updates
```
`IsStillZombieChannel`?

Roasbeef commented on routing/router.go:
```diff
@@ -885,39 +910,27 @@ func (r *ChannelRouter) pruneZombieChans() error {
+		// If either edge hasn't been updated for a period of
+		// chanExpiry, then we'll mark the channel itself as eligible
+		// for graph pruning.
+		if !r.IsZombieChannel(e1Time, e2Time) {
```
👍

Roasbeef commented on routing/router.go:
```diff
@@ -885,39 +910,27 @@ func (r *ChannelRouter) pruneZombieChans() error {
 				info.NodeKey1Bytes, info.ChannelID)
 		}
-		if e2Zombie {
+		if e2 == nil || time.Since(e2.LastUpdate) >= chanExpiry {
```
If we want to fully consolidate this code block, then we can have an internal version of `IsZombieChannel` that returns a bool for `e1` and `e2`. Then we check the conjunction below.

Roasbeef commented on channeldb/graph.go:
```diff
@@ -2092,13 +2095,19 @@ func (c *ChannelGraph) FilterKnownChanIDs(chanIDs []uint64) ([]uint64, error) {
+				isZombie, _, _ := isZombieEdge(
+					zombieIndex, scid,
+				)
+				if isZombie && isZombieChan(
```
Ah yeah, this makes sense, so scratch the idea about the cache I had. I had envisioned a look up here each time to double check our timestamp with theirs. 

Thinking about it a bit more, perhaps we should do some sort of sanity check? Otherwise, they can send bogus timestamps and make us request everything all over again.

morehouse replied:
How would such a sanity check work?  I imagine a malicious peer could trivially defeat any check by setting timestamps to a recent time.

Roasbeef reviewed (changes requested):
Nice work! Main comment re if we should attempt to sanity check the timestamps we get from the remote party. 

One way to test this out in the wild would be: create a new node on mainnet, no channels, set the zombie prune horizon to something short like 6 hrs, run this patch, and observe the zombie resurrection happen or not. We might want to add some extra logging statements wehn we realize we need to go fetch a zombie chan due to getting an update.

lightninglabs-deploy commented:
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ellemouton, remember to re-request review from reviewers when ready

Crypt-iQ reviewed (commented):
Still in the process of reviewing, but I think the `chunkSize` needs to be updated if timestamps are sent because instead of room for ~8000 scid's, there's only room for ~4000 because of the 2 4-byte timestamps

ellemouton commented:
> Still in the process of reviewing, but I think the chunkSize needs to be updated if timestamps are sent because instead of room for ~8000 scid's, there's only room for ~4000 because of the 2 4-byte timestamps

Great catch - thanks! Addressed 👍 

> Main comment re if we should attempt to sanity check the timestamps we get from the remote party.

👍  for now i've added a place holder that currently only checks that the times are not in the future. Will ponder more over if more can be done & if it is worth doing something more complicated like checking that what they end up sending matches the timestamps they claimed to have. 

> One way to test this out in the wild would be: create a new node on mainnet, no channels, set the zombie prune horizon to something short like 6 hrs, run this patch, and observe the zombie resurrection happen or not. We might want to add some extra logging statements wehn we realize we need to go fetch a zombie chan due to getting an update.

Good idea 👍  I'll do this next

ellemouton commented:
Did the following manual testing:

### Confirmed issue:

1. Set up a regtest env where there are a bunch of chans but main node, alice, has no chans but learns of all the other channels.
2. Set alice's `ChannelPruneExpiry` and `GraphPruneInterval` to low values so that the node quickly marks all the other channels as zombies. (a quick way to verify this is to do a `describegraph` and see that there are no edges. 
3. Shutdown alice
4. now perform some channel updates on the existing channels so that there are some ChannelUpdates that alice misses while the node is offline. 
5. Start alice , connect alice to a node in the network & see that Alice ignores all the SCIDs in the lnwire.ReplyChannelRange message due to them all being marked as zombies.  (`describegraph` still empty). 

### PR fixes issue:
Restart alice with this PR and perform the above steps. At step 5, alice will mark the channels as alive and will re-add them to her graph.

saubyk commented:
> ...
> 5. Start alice , connect alice to a node in the network & see that Alice ignores all the SCIDs in the lnwire.ReplyChannelRange message due to them all being marked as zombies.  (`describegraph` still empty).
> 
> ### PR fixes issue:
> Restart alice with this PR and perform the above steps. At step 5, alice will mark the channels as alive and will re-add them to her graph.

Great explanation of the test 🤩 

Concept ACK

Crypt-iQ commented on discovery/syncer.go:
```diff
@@ -906,6 +935,17 @@ func (g *GossipSyncer) processChanRangeReply(msg *lnwire.ReplyChannelRange) erro
+	//  the full messages that they send?
+
+	// A timestamp in the future is not considered sane.
+	return !t.After(time.Now())
```
I'm not sure if this works, what if there is clock drift? I'm also not sure how useful this sanity check is -- what's the worst an attacker can do without the sanity check?

morehouse replied:
AFAICT, the worst an attacker can do is trick us into unnecessarily requesting *all* channel updates within the range.  Considering that would put some load on the attacker themselves to send those updates, I don't think it's much of an attack.

Either way, this simple sanity check can be trivially defeated by using `time.Now()` as the timestamp.

ellemouton replied:
> Considering that would put some load on the attacker themselves to send those updates, I don't think it's much of an attack.

good point. 

removed 👍

morehouse commented on channeldb/graph.go:
```diff
@@ -2118,13 +2120,37 @@ func (c *ChannelGraph) FilterKnownChanIDs(chanIDs []uint64) ([]uint64, error) {
+				// and we let it be added to the set of IDs to
+				// query our peer for.
+				case isZombie && !isStillZombie:
+					err := c.markEdgeLive(tx, scid)
```
Isn't marking the edge live here a little premature?  We haven't requested and checked the new channel update yet.

Once we do process the channel update, the existing code should already mark the edge live.

Crypt-iQ replied:
Trying to think about the case where a random peer can make us remove arbitrary entries from our zombie index bucket w/o having more recent updates for these channels. It doesn't seem like an issue as if we ever received the updates after removing them from the zombie index, we'd just go through the validation pipeline again, but figured I'd point it out

morehouse replied:
If truly dead edges are marked live, could that degrade payment success rates?

morehouse replied:
This code from `MarkEdgeLive` concerns me:

https://github.com/lightningnetwork/lnd/blob/f005b248ced9bcd2316c0cbe7c9c25298796513c/channeldb/graph.go#L3529-L3542

It looks like we're activating the edge for pathfinding.

ellemouton replied:
`c.FetchChanInfos` here will return an empty list though since we would have deleted the edge when we pruned this channel via [`delChannelEdge`](https://github.com/lightningnetwork/lnd/blob/f005b248ced9bcd2316c0cbe7c9c25298796513c/channeldb/graph.go#L2379)

 which first [deletes the edge](https://github.com/lightningnetwork/lnd/blob/f005b248ced9bcd2316c0cbe7c9c25298796513c/channeldb/graph.go#L2435) 

before [marking it as a zombie](https://github.com/lightningnetwork/lnd/blob/master/channeldb/graph.go#L2458)

ellemouton replied:
so `MarkEdgeLive` will only load the channel edge info back into our graph cache if we still have the channel edge on disk which we wont in this case

morehouse replied:
Makes sense, thanks for explaining!

ellemouton commented on discovery/syncer.go:
```diff
@@ -944,6 +944,7 @@ func (g *GossipSyncer) genChanRangeQuery(
 		ChainHash:        g.cfg.chainHash,
 		FirstBlockHeight: startHeight,
 		NumBlocks:        numBlocks,
+		QueryOptions:     lnwire.NewTimestampQueryOption(),
```
discussed off-line: will update tomorrow to let it switch off both ways

ellemouton replied:
updated to disable both ways if set!

ellemouton commented on lnwire/reply_channel_range.go:
```diff
@@ -103,7 +130,66 @@ func (c *ReplyChannelRange) Encode(w *bytes.Buffer, pver uint32) error {
 	}
 
-	err := encodeShortChanIDs(w, c.EncodingType, c.ShortChanIDs, c.noSort)
+	timestamps := c.Timestamps
```
it's used lower down where c.Timestamps gets reconstructed. Moved it closer to there 👍

ellemouton commented on lnwire/reply_channel_range.go:
```diff
@@ -103,7 +130,66 @@ func (c *ReplyChannelRange) Encode(w *bytes.Buffer, pver uint32) error {
+					"pair for each of the given SCIDs")
+			}
+
+			// Create a map from SCID value to the original index of
+			// the SCID in the unsorted list.
+			scidPreSortIndex = make(
+				map[uint64]int, len(c.ShortChanIDs),
+			)
+			for i, scid := range c.ShortChanIDs {
+				scidPreSortIndex[scid.ToUint64()] = i
+			}
```
i did originally try that but it doesnt work since the timestamp list gets changed throughout the processing of `.Slice` and so as soon as one time gets swapped out, the `i`, and `j` indices become inconsistent.

ellemouton commented on lnwire/timestamps.go:
```diff
@@ -0,0 +1,137 @@
+				Timestamp1: t1,
+				Timestamp2: t2,
+			}
+		}
```
nice catch

morehouse reviewed (approved):
LGTM

lightninglabs-deploy commented:
@roasbeef: review reminder

Roasbeef reviewed (approved):
LGTM 🌲
